File: 4-Median of Two Sorted Arrays.py
class Solution:
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        x, y, i = 0, 0, 0
        big = []
        while x < len(nums1) and y < len(nums2):
            if nums1[x] <= nums2[y]:
                big.append(nums1[x])
                x, i = x + 1, i + 1
            else:
                big.append(nums2[y])
                y, i = y + 1, i + 1
        while x < len(nums1):
            big.append(nums1[x])
            i, x = i + 1, x + 1
        while y < len(nums2):
            big.append(nums2[y])
            i, y = i + 1, y + 1
        if len(big) % 2 == 1:
            return big[int(len(big) / 2)]
        else:
            return (big[int(len(big) / 2)] + big[int(len(big) / 2 - 1)]) / 2


'''
常规操作
将两个数组排序合并
然后计算合并后数组的中位数
'''

File: C++/4-Median_of_Two_Sorted_Arrays.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        vector<int> tmp;
        int i = 0, j = 0, k = 0;
        while (i < nums1.size() && j < nums2.size()) {
            if (nums1[i] < nums2[j]) {
                tmp.push_back(nums1[i]);
                i++;
                k++;
            } else {
                tmp.push_back(nums2[j]);
                j++;
                k++;
            }
        }
        while (i < nums1.size()) {
            tmp.push_back(nums1[i]);
            i++;
            k++;
        }
        while (j < nums2.size()) {
            tmp.push_back(nums2[j]);
            j++;
            k++;
        }
        if (k % 2 == 0)
            return (tmp[k / 2 - 1] + tmp[k / 2]) / 2.0;
        else
            return tmp[k / 2];
    }
};


class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        double res = 0;
        int i = 0, j = 0, k = 0;
        int stop = (nums1.size() + nums2.size()) / 2 - 1;  // the first stop index
        bool len_is_even = (nums1.size() + nums2.size()) % 2 == 0;
        while ((i < nums1.size() || j < nums2.size())) {
            if (i < nums1.size() && j < nums1.size()) {
                if (nums1[i] < nums2[j]) {
                    ++i; ++k;
                } else {
                    ++j; ++k;
                }
            } else if (i < nums1.size()) {
                ++i; ++k;
            } else {
                ++j; ++k;
            }
            if (k == stop || k == stop + 1) {
                res += max(nums1[i - 1], nums2[j - 1]);
                if (len_is_even && k == stop + 1) {
                    res += max(nums1[i - 1], nums2[j - 1]);
                    return res / 2.0;
                } else
                    return res;
            }
        }
    }
};